### PR TITLE
fix(router): resolve KiCad layer names in net-class-map layer lists

### LINKED ISSUE #4587: route: grid engine crashes on layer-name strings in net-class-map (int('In1.Cu')) — lattice accepts both formats
### Repro

kicad-tools main @ 45744d0a, softstart rev-C (160×100mm 4-layer, 97 nets), native backend built.

```
kct route board.kicad_pcb --route-engine grid --strategy negotiated --layers 4 \
  --grid auto --max-cells 20000000 --net-class-map net_class_map.json \
  --manufacturer jlcpcb --voltage-map vmap.json
```

Net-class-map entries carry layer intent as KiCad layer names, e.g.:

```json
"/AC_LINE": { "preferred_layers": ["F.Cu", "B.Cu"], "avoid_layers": ["In1.Cu", "In2.Cu"] }
```

### Result

Detailed routing dies on the first net:

```
--- Phase 2: Detailed Routing ---
  [  0.0%] Routing /PGND... (0.0s)
Error during routing: invalid literal for int() with base 10: 'In1.Cu'
```

Exit 1 after the full escape/global phases (~minutes wasted); partial snapshot saved.

### Expected

The **lattice** engine consumes the same map without complaint (it also accepts numeric indices — our map historically mixes `"preferred_layers": [3]` and `["F.Cu","B.Cu"]`, both lattice-accepted). The two engines should share one normalization: accept both layer names and indices, coerced at map load, not deep inside detailed routing.

### Why it matters now

Phase 2 pairwise HV search-time avoidance (#4511 / #4524) lives in the C++ grid pathfinder — the lattice engine bypasses it (routes exit 0, then fail the pairwise census; 129 board-level fails on this board). So an HV board is steered to `--route-engine grid` for #4431 enforcement, and immediately trips this crash if its map uses layer names.

### Workaround

Pre-translate names→indices (F.Cu→0, In1.Cu→1, In2.Cu→2, B.Cu→3) in a derived map copy; grid engine then proceeds.

---

### Root cause (confirmed against `origin/main` @ 092611b0)

There is exactly **one** integer coercion of layer values anywhere in the repo, and it is the crash site:

`src/kicad_tools/router/rules.py:1160-1164` — `NetClassRouting.hard_avoided_layer_indices()`

```python
if not self.avoid_layers:
    return frozenset()
if self.target_ampacity is not None or strict_layers:
    return frozenset(int(layer) for layer in self.avoid_layers)   # <-- int('In1.Cu')
return frozenset()
```

Verified exhaustive: `grep -rn "int(layer" src/` returns only `rules.py:1163`, `drc/different_net_short.py:102` (unrelated — parses `In<k>.Cu` ordinals), and `router/io.py:2940` (unrelated — parses the PCB `(layers ...)` block).

The two callers are the two grid backends:
- `src/kicad_tools/router/pathfinder.py:2296` (`_hard_avoided_layers`, Python backend)
- `src/kicad_tools/router/cpp_backend.py:1348` (native backend)

**Trigger condition** (explains why this is not a universal crash): `hard_avoided_layer_indices()` short-circuits to an empty set unless the class declares `target_ampacity` **or** `DesignRules.strict_layers` is on. The reporter's `/PGND` class is ampacity-bearing, so the hardening branch runs and `int('In1.Cu')` raises. A name-bearing map with no ampacity and no `--strict-layers` does **not** crash — it silently does nothing (see below).

### Correction to the issue's premise: the lattice engine does not "accept" layer names — it ignores the fields entirely

Verified: `grep -rn "avoid_layers|preferred_layers" src/kicad_tools/router/lattice/ src/kicad_tools/router/mesh/` returns **zero** hits. The lattice pathfinder reads only `trace_width`, `clearance`, and `neck_trace_width` off the net class (`src/kicad_tools/router/lattice/pathfinder.py:447-457, 624-635`). So "lattice accepts both formats" is really "lattice never looks at either format". **Do not** copy a resolution strategy out of the lattice engine — there isn't one. The canonical, stack-aware name→grid-index lookup that *does* exist is `LayerStack.get_layer_by_name(name).index` (`src/kicad_tools/router/layers.py:113-118`), which is the same table `RoutingGrid` builds its own `_layer_to_index` from (`src/kicad_tools/router/grid.py:729-736`). Reuse that; do not invent a second parser.

### The silent sibling defect (fix in the same pass)

`preferred_layers` and the **soft** `avoid_layers` cost bias are plain membership tests against an `int` grid index:

`src/kicad_tools/router/pathfinder.py:2266-2274`
```python
if net_class.preferred_layers is not None:
    if layer in net_class.preferred_layers:      # int in ["F.Cu", "B.Cu"] -> always False
        return 0.5
if net_class.avoid_layers is not None:
    if layer in net_class.avoid_layers:          # int in ["In1.Cu", ...] -> always False
        return net_class.layer_cost_multiplier
```

So a name-bearing map today is **silently inert** for layer preference/bias and **crashes** only on the hard-avoid path. Both halves are the same root cause (names never normalized to grid indices) and both are fixed by normalizing once at deserialization. The dataclass contract already declares the intent: `rules.py:843-844` annotates both fields as `list[int] | None`.

### Do NOT use `CopperLayer.from_kicad_name()` — it is the wrong index space

`core/types.py` `CopperLayer` is a fixed 6-slot enum (`F_CU=0, IN1=1, IN2=2, IN3=3, IN4=4, B_CU=5`). On a 4-layer board the grid index of `B.Cu` is **3**, not 5. Resolving `B.Cu` through `CopperLayer` would silently place an HV net's hard-avoid on a non-existent layer index — a wrong-but-quiet failure strictly worse than today's crash. `B.Cu` is only resolvable against the board's actual layer count. `F.Cu` → `0` and `In<k>.Cu` → `k` are unambiguous in every KiCad stackup.

### Recommended approach

**Part 1 — `src/kicad_tools/router/rules.py` (required)**

1. Add a module-level resolver, e.g. `resolve_layer_index(value, layer_stack=None) -> int`:
   - `int` → returned as-is (existing maps stay byte-identical);
   - `str` of digits (`"3"`) → `int(value)`;
   - copper layer name with a `layer_stack` supplied → `layer_stack.get_layer_by_name(name).index` (canonical path);
   - copper layer name with no stack → `F.Cu` → `0`, `In<k>.Cu` → `k`; `B.Cu` raises `ValueError` naming the value and stating that a layer count is required;
   - anything else → `ValueError` naming the offending token and listing the accepted forms.
2. Normalize **both** `preferred_layers` and `avoid_layers` through it inside `NetClassRouting.from_dict` (`rules.py:1327-1362`), adding an optional `layer_stack=None` kwarg. The stored fields then always satisfy their `list[int] | None` annotation.
3. Forward the same optional kwarg from `net_class_map_from_dict` (`rules.py:1559-1596`). Default `None` keeps all six existing callers (`cli/route_cmd.py`, `cli/check_cmd.py`, `cli/zones_cmd.py`, `cli/commands/creepage.py`, `audit/auditor.py` ×2) working unchanged for integer maps.
4. Replace the bare `int(layer)` at `rules.py:1163` with the shared resolver so any residual string raises an actionable error naming the net class and the token instead of `invalid literal for int() with base 10: 'In1.Cu'`. **Fail loud — do not silently skip an unresolvable entry**: a silently-dropped hard avoid puts an ampacity-bearing HV net onto a 0.5 oz inner plane, which is exactly the failure #4433 exists to prevent.

**Part 2 — `src/kicad_tools/cli/route_cmd.py`, strictly inside lines 11490-11567 (required)**

The `--net-class-map` preload block is at `route_cmd.py:11498-11524`. Resolve the layer stack there and pass it into `net_class_map_from_dict(...)` so names are gone before any engine touches the map. Mirror the existing selection logic verbatim (`route_cmd.py:5543-5555`, duplicated at `12018-12028`):

```python
if args.layers == "auto":
    layer_stack = detect_layer_stack(pcb_path.read_text())   # router.io
else:
    layer_stack = {"2": LayerStack.two_layer(), "4": ..., "4-sig": ..., "4-all": ..., "6": ...}[args.layers]
```

`pcb_path` is already in scope in this block (used at `route_cmd.py:11497` by `_offboard_preflight`). The existing `except (TypeError, ValueError)` at `route_cmd.py:11522-11524` already converts a bad name into `Error: invalid net-class-map structure: ...` + exit 1 — **before** the escape/global phases burn minutes, which is the reporter's other complaint.

**Rejected alternative**: resolving lazily inside `pathfinder.py` / `cpp_backend.py` at their `hard_avoided_layer_indices` callsites. It duplicates the logic across two backends, leaves the soft-cost path in `_get_layer_preference_cost` still broken for names, and pushes the failure back to Phase 2 instead of preload.

### Scope reservation — coordination with in-flight #4588

#4588 is **being built right now** in a parallel worktree and also touches `src/kicad_tools/cli/route_cmd.py`. The regions are reserved and disjoint; keep them that way so both PRs stay mergeable:

| Issue | `route_cmd.py` regions | Other files |
|---|---|---|
| **#4588** (in flight) | new audit helper near `:3808-3860`; success-predicate block near `:13740-13790` | `src/kicad_tools/router/pairwise_clearance.py` |
| **#4587** (this issue) | **only** the `--net-class-map` preload block, `:11490-11567` | `src/kicad_tools/router/rules.py` |

Hard rules for this issue's PR:
- All new `route_cmd.py` code must live **inline inside `:11490-11567`**. Do **not** add a new module-level helper elsewhere in `route_cmd.py` — it risks landing in #4588's hunks.
- Do **not** touch `_apply_net_class_map_sidecar` / `_apply_pairwise_clearance` (`route_cmd.py:3755-3860`) even though the sidecar merge site looks like a natural home for the resolution. That range is inside #4588's blast radius.
- Do **not** touch the route success predicate or `src/kicad_tools/router/pairwise_clearance.py`.

### Acceptance Criteria

- [ ] `kct route --route-engine grid --net-class-map <map>` completes without `invalid literal for int() with base 10: 'In1.Cu'` when the map declares `avoid_layers` / `preferred_layers` as KiCad layer names, including on an ampacity-bearing class (the crash's trigger condition).
- [ ] A map that mixes formats within one file (`"preferred_layers": [3]` alongside `"avoid_layers": ["In1.Cu","In2.Cu"]`) loads and routes.
- [ ] `B.Cu` resolves to the board's **last** grid index (`3` on a 4-layer stack, `1` on 2-layer, `5` on 6-layer) — not to `CopperLayer.B_CU == 5`.
- [ ] `F.Cu` → `0` and `In<k>.Cu` → `k` on every stack.
- [ ] Layer names now take effect on the **soft** paths too: `preferred_layers: ["F.Cu"]` produces the same `_get_layer_preference_cost` discount as `preferred_layers: [0]` (previously silently inert).
- [ ] An unrecognized layer token (e.g. `"Top"`, `"F.Silkscreen"`) fails at **preload** with `Error: invalid net-class-map structure: ...` naming the offending value and the owning net-class key, exit 1, before any escape/global routing runs.
- [ ] `hard_avoided_layer_indices()` raises an actionable error (naming the class and token) rather than `int()`'s opaque `ValueError` if it is ever handed an unresolved string. It must **not** silently drop the entry.
- [ ] Integer-valued maps are byte-identical in behavior: an existing routed board with an integer `avoid_layers` map produces the same route output as before the change (no-op regression guard).
- [ ] `to_dict()` / `from_dict()` round-trip drift guard in `tests/test_net_class_serialization.py` still passes (names normalize to ints on the way in; `to_dict` keeps emitting ints, so the round trip is idempotent).
- [ ] The other five `net_class_map_from_dict` callers (`check_cmd.py:1451`, `zones_cmd.py:672`, `commands/creepage.py:175`, `auditor.py:898`, `auditor.py:1237`) compile and pass unchanged with the new optional kwarg defaulted.
- [ ] **Scope guard**: the diff touches only `src/kicad_tools/router/rules.py`, `src/kicad_tools/cli/route_cmd.py` within `:11490-11567`, and test files. Zero hunks in `src/kicad_tools/router/pairwise_clearance.py` or in `route_cmd.py:3755-3860` / `:13740-13790` (reserved for in-flight #4588).

## Affected Files

- `src/kicad_tools/router/rules.py` — add the shared layer-index resolver; normalize `preferred_layers` / `avoid_layers` in `NetClassRouting.from_dict` (`:1327-1362`); forward the optional stack kwarg from `net_class_map_from_dict` (`:1559-1596`); replace the bare `int(layer)` at `:1163` with the resolver + actionable error.
- `src/kicad_tools/cli/route_cmd.py` — **only** the `--net-class-map` preload block at `:11490-11567`: resolve the layer stack (mirroring `:5543-5555`) and pass it to `net_class_map_from_dict`. No edits anywhere else in this file.
- `tests/test_net_class_serialization.py` — add name-form round-trip cases beside the existing `test_layer_preferences_roundtrip` (`:70-81`); assert names normalize to ints and that the existing int round trip is unchanged.
- `tests/test_cli_route_net_class_map.py` — add a CLI-level regression: a sidecar with layer-name `avoid_layers` on an ampacity-bearing class must not raise `invalid literal for int()`; an unrecognized token must exit 1 at preload with the structured message.

Reference-only (read, do not modify): `src/kicad_tools/router/layers.py` (`LayerStack.get_layer_by_name`), `src/kicad_tools/router/io.py` (`detect_layer_stack`), `src/kicad_tools/router/pathfinder.py:2266-2296`, `src/kicad_tools/router/cpp_backend.py:1348`, `src/kicad_tools/core/types.py` (`CopperLayer` — the wrong index space, see above).

## Test Plan

- [ ] Manual verification: build a 4-layer fixture sidecar with `{"preferred_layers": ["F.Cu","B.Cu"], "avoid_layers": ["In1.Cu","In2.Cu"], "target_ampacity": 15.0}` and run `kct route <board> --route-engine grid --layers 4 --net-class-map <map>`. Pre-fix: `Error during routing: invalid literal for int() with base 10: 'In1.Cu'`. Post-fix: routing proceeds and the class's traces carry zero In1/In2 copper.
- [ ] Manual verification: repeat with `--layers auto` on the same board to confirm the auto-detected stack resolves `B.Cu` to index 3.
- [ ] Automated tests: `tests/test_net_class_serialization.py` — name→index normalization for `F.Cu` / `In1.Cu` / `In2.Cu` / `B.Cu` across 2-, 4-, and 6-layer stacks; mixed int/name lists; idempotence (`from_dict(to_dict(x)) == x`).
- [ ] Automated tests: `tests/test_cli_route_net_class_map.py` — preload rejects an unrecognized token with exit 1 and a message naming the token and the net-class key; an ampacity-bearing name-based class routes without raising.
- [ ] Regression: `uv run pytest tests/test_router_core.py -k avoid_layers` (the #4433 hard-avoid suite at `:4897-5050`) stays green — it exercises the integer path in both backends and is the no-op guard for existing maps.
- [ ] Edge cases: empty list; `null`; a name for a layer not present in the stack (e.g. `In3.Cu` on a 4-layer board) — must surface loudly, not silently produce an inert index; `"3"` as a numeric string.
- [ ] Full check: `uv run ruff check . && uv run mypy src/ && uv run pytest tests/test_net_class_serialization.py tests/test_cli_route_net_class_map.py tests/test_cli_check_net_class_map.py`.

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

`kct route --route-engine grid --net-class-map <map>` died on the first net with
`Error during routing: invalid literal for int() with base 10: 'In1.Cu'` whenever the
sidecar spelled `avoid_layers` / `preferred_layers` as KiCad layer names on an
ampacity-bearing class — after the escape and global phases had already burned minutes.

There was exactly one integer coercion of layer values in the repo, and it was the crash
site: `NetClassRouting.hard_avoided_layer_indices()` (`router/rules.py:1163`,
`frozenset(int(layer) for layer in self.avoid_layers)`), reached from both grid backends
(`router/pathfinder.py:2296`, `router/cpp_backend.py:1348`).

The same root cause has a **silent sibling defect**: the soft layer-preference / avoid
cost bias (`router/pathfinder.py:2266-2274`) is a plain `in` membership test against an
int grid index, so layer names there never crashed and never matched — silently inert.

Both halves are fixed by normalizing layer values **once, at deserialization**, so no
engine ever sees a name.

Note on the issue title's premise: the lattice engine does **not** accept both formats —
it never reads either field (`grep` over `router/lattice/` + `router/mesh/` returns zero
hits for `avoid_layers|preferred_layers`). There was no lattice parser to reuse.

## Changes

`src/kicad_tools/router/rules.py`
- New `resolve_layer_index(value, layer_stack=None, *, context="")` + `LayerResolutionError`
  (a `ValueError`, so every existing `except (TypeError, ValueError)` guard keeps catching it).
  Accepts an `int` (returned unchanged), a numeric string (`"3"`), or a KiCad copper layer
  name. Names resolve through `LayerStack.get_layer_by_name(name).index` — the same table
  `RoutingGrid._layer_to_index` is built from — when a stack is supplied; without a stack only
  the stack-independent forms resolve (`F.Cu` -> 0, `In<k>.Cu` -> k) and `B.Cu` raises.
- **`CopperLayer.from_kicad_name()` is deliberately NOT used.** That fixed 6-slot enum has
  `B_CU == 5`, but `B.Cu` is grid index **3** on a 4-layer board — using it would convert a
  loud crash into a silent wrong answer (hard-blocking a non-existent layer index).
- `NetClassRouting.from_dict(data, layer_stack=None)` normalizes **both** layer lists;
  `net_class_map_from_dict(data, layer_stack=None)` forwards the kwarg and re-raises layer
  failures tagged with the owning sidecar key.
- The bare `int(layer)` at the crash site now goes through the resolver, so a residual string
  raises an actionable error naming the class and token instead of `int()`'s opaque message —
  and is never silently dropped (a dropped hard-avoid puts an ampacity-bearing net back on a
  thin inner plane, exactly what #4433 exists to prevent).

`src/kicad_tools/cli/route_cmd.py` — **only** the `--net-class-map` preload block
- Resolves the layer stack there (`--layers auto` detection or the explicit preset, mirroring
  the existing `--layers` handling) and passes it to `net_class_map_from_dict`. Stack resolution
  is best-effort; failure degrades to "no stack" (names still resolve except `B.Cu`, which then
  fails loud) rather than aborting. The existing `except (TypeError, ValueError)` already turns a
  bad token into `Error: invalid net-class-map structure: ...` + exit 1 at preload.

Tests: `tests/test_net_class_serialization.py` (+21) and `tests/test_cli_route_net_class_map.py` (+7).

## Acceptance Criteria Verification

Verified against **board-02 (charlieplex)** forced to `--layers 4`, with a sidecar declaring
`preferred_layers: ["F.Cu","B.Cu"]`, `avoid_layers: ["In1.Cu","In2.Cu"]`, `target_ampacity: 15.0`.

**Pre-fix (`origin/main` checkout, native backend, identical command):**

```
--- Iteration 0: Initial routing with sharing ---
  [  0.0%] Routing net 1/8: LINE_C... (0.0s)
  ...
  [ 62.5%] Routing net 6/8: LINE_B... (0.4s)
Error during routing: invalid literal for int() with base 10: 'In1.Cu'
  Raw partial snapshot saved to: prefix_charlie_partial.kicad_pcb
```

**Post-fix (this branch, same command):** all 8 nets route to completion, no `int()` error.
Segment-per-layer census of the output board — the three name-declared HV classes carry
**zero** In1/In2 copper while unconstrained nets use the inner layers freely:

```
HV-class (avoid In1.Cu/In2.Cu) segment layers:
   LINE_A   B.Cu     2
   LINE_A   F.Cu     2
   LINE_B   F.Cu     7
   NODE_A   B.Cu     11
   NODE_A   F.Cu     26

Non-HV nets, for contrast:
   LINE_D   In1.Cu   8
   NODE_B   In1.Cu   25
   NODE_C   In1.Cu   4
   NODE_C   In2.Cu   1
   NODE_D   In1.Cu   100
```

- [x] No `invalid literal for int() with base 10: 'In1.Cu'` on an ampacity-bearing class — see above.
- [x] Mixed formats in one file (`"preferred_layers": [3]` + `"avoid_layers": ["In1.Cu","In2.Cu"]`)
      load and route — covered by `test_mixed_int_and_name_forms_in_one_file` and by the board run.
- [x] `B.Cu` resolves to the board's **last** grid index, not `CopperLayer.B_CU == 5`. Real CLI run
      with `--layers auto` on a 4-layer board:

      ```
      === PROBE --layers auto ===
      detected stack: 4-Layer SIG-SIG-GND-PWR (4 copper layers)
      stack names   : ['F.Cu', 'In1.Cu', 'In2.Cu', 'B.Cu']
        LINE_A: preferred_layers=[0, 3] avoid_layers=[1, 2]
      CopperLayer.B_CU.value (the WRONG index space) = 5
      ```

      Parametrized over 2-/4-/6-layer stacks (1 / 3 / 5) in
      `test_b_cu_resolves_to_the_stacks_last_index`.
- [x] `F.Cu` -> 0 and `In<k>.Cu` -> k on every stack.
- [x] Layer names now take effect on the **soft** path:
      `test_soft_preference_cost_matches_the_integer_form` asserts
      `Router._get_layer_preference_cost` is identical for `["F.Cu"]` and `[0]` (0.5 discount,
      previously the neutral 1.0).
- [x] An unrecognized token (`"Top"`, `"F.Silkscreen"`) fails at **preload**, exit 1, naming the
      value and the owning net-class key, before any routing —
      `test_unrecognized_token_exits_1_at_preload`.
- [x] `hard_avoided_layer_indices()` raises an actionable, class-and-token-naming error for a
      residual string and does not drop the entry —
      `test_hard_avoid_raises_actionable_error_for_residual_string`.
- [x] Integer-valued maps behave identically (ints pass through untouched) —
      `test_integer_form_is_unchanged`, `test_integer_sidecar_is_unchanged`, and
      `pytest tests/test_router_core.py -k avoid_layers` (the #4433 suite) green.
- [x] `to_dict()` / `from_dict()` round-trip drift guard still passes; names normalize on the way
      in, `to_dict` keeps emitting ints, round trip idempotent —
      `test_roundtrip_is_idempotent_for_name_form` + the existing `TestDriftPrevention` suite.
- [x] The other five `net_class_map_from_dict` callers (`check_cmd.py`, `zones_cmd.py`,
      `commands/creepage.py`, `auditor.py` x2) are untouched and pass with the kwarg defaulted —
      `tests/test_cli_check_net_class_map.py`, `tests/test_audit.py`, `tests/creepage` green.
- [x] **Scope guard**: the diff touches only `router/rules.py`, `route_cmd.py` inside the reserved
      preload window, and two test files. `git diff -U0` hunk headers for `route_cmd.py` are
      `@@ -11519,0 +11520,15 @@` and `@@ -11521 +11536,19 @@` — both inside `:11490-11567`.
      **Zero** hunks in `router/pairwise_clearance.py`, `route_cmd.py:3755-3860`
      (`_apply_net_class_map_sidecar` / `_apply_pairwise_clearance`), or `route_cmd.py:13740-13790`
      (the success predicate) — all reserved for in-flight #4588.

## Test Plan

- `pytest tests/test_net_class_serialization.py tests/test_cli_route_net_class_map.py tests/test_cli_check_net_class_map.py` — **103 passed**
- `pytest tests/test_router_core.py -k "avoid_layers or layer_pref"` — **6 passed** (the #4433 no-op guard)
- `pytest tests/test_router_core.py tests/test_audit.py tests/creepage` — **571 passed**
- `pytest tests/ -k "net_class or netclass or rules or layer"` — **1687 passed, 4 skipped**
- `ruff check .` — All checks passed; `ruff format --check` clean on all four changed files
- `scripts/ci/check_mypy_baseline.py` — "all within the baseline (1474 allowed). No new type errors."
  (Baseline file intentionally not updated: the one no-longer-occurring line is the env-dependent
  nanobind `import-not-found`, which only disappears in a natively-built worktree.)
- Manual: pre-fix crash + post-fix completion on board-02 at `--layers 4`, and `--layers auto`
  stack detection on a 4-layer board (both pasted above). C++ backend built and active
  (`kct build-native --check` -> `available (version 1.0.0)`).

Closes #4587

### CODE REVIEW

**Comment by rjwalters** [comment]
## APPROVED

Reviewed at head `b90d32b8` (integrated with `main` @ `5cc7690a`). All 16 CI checks pass, merge state CLEAN. Verified independently rather than taking the Builder's evidence on trust.

### 1. `B.Cu` resolves to 3, not 5 — the acceptance test

Ran the resolver against every stack preset the CLI can select:

```
2      names=['F.Cu', 'B.Cu']                                     B.Cu -> 1
4      names=['F.Cu', 'In1.Cu', 'In2.Cu', 'B.Cu']                 B.Cu -> 3
4-sig  names=['F.Cu', 'In1.Cu', 'In2.Cu', 'B.Cu']                 B.Cu -> 3
4-all  names=['F.Cu', 'In1.Cu', 'In2.Cu', 'B.Cu']                 B.Cu -> 3
6      names=['F.Cu','In1.Cu','In2.Cu','In3.Cu','In4.Cu','B.Cu']  B.Cu -> 5
CopperLayer.B_CU.value (the WRONG index space) = 5
CopperLayer.from_kicad_name('B.Cu').value       = 5
```

`B.Cu` tracks `num_layers - 1` on every stack. The `CopperLayer` trap is confirmed unreachable: the only occurrence of `CopperLayer` anywhere in `router/rules.py` is inside a docstring (line 769) explaining why it is not used. Resolution goes exclusively through `LayerStack.get_layer_by_name(token).index`, and when no stack is supplied only the stack-independent forms resolve (`F.Cu` -> 0, `In<k>.Cu` -> k) while `B.Cu` raises with an actionable message. That is the correct call — a guess there would be a silent wrong answer.

The CLI plumbing was verified too, not just the resolver: `--layers 2` yields `avoid_layers == [1]` and `--layers 6` yields `[5]` for the same `["B.Cu"]` sidecar, and `--layers auto` detects 4 copper layers from the board file.

### 2. The hard-avoid is genuinely in force — not a silent no-op

Crash reproduced first on the same tree with only the two source files reverted to `origin/main`, board-02 at `--layers 4` with an ampacity-bearing name-declared class:

```
[ 62.5%] Routing net 6/8: LINE_B... (0.7s)
Error during routing: invalid literal for int() with base 10: 'In1.Cu'
```

With the PR sources restored, the same command routes 8/8. Segment census of the written board (LINE_A/B/C declared `avoid_layers: ["In1.Cu","In2.Cu"]` + `target_ampacity`):

```
name-declared HV classes on In1/In2 : 0 segments
unconstrained nets on In1/In2       : 291 segments   (NODE_D In2.Cu 187, NODE_C In1.Cu 66, ...)
```

Because the three HV nets in that run are small and might plausibly have stayed on outer layers anyway, I added a **positive control**: put the two nets that naturally live on the inner layers under a name-declared hard avoid and re-route.

```
                          F.Cu   In1.Cu  In2.Cu  B.Cu
NODE_C baseline             69      66      15      0
NODE_C avoid ["In1.Cu"]     27       0       0      8
NODE_D baseline             30       0     187      0
NODE_D avoid ["In2.Cu"]     26     105       0      0
```

NODE_D's 187 In2.Cu segments go to zero and relocate to In1.Cu; NODE_C's 66 In1.Cu segments go to zero. The name resolved to the right index and was hard-blocked end to end through the native backend. This is not an empty-set no-op.

### 3. The silent sibling defect is closed

`_get_layer_preference_cost` is unchanged, which is right — it was never broken, it was being fed names. With normalization at deserialization it now matches:

```
by_name  = from_dict({"preferred_layers": ["F.Cu"], "avoid_layers": ["In1.Cu"]})
by_index = from_dict({"preferred_layers": [0],      "avoid_layers": [1]})
cost(0) == 0.5 for both; cost(1) == layer_cost_multiplier for both
```

`test_soft_preference_cost_matches_the_integer_form` covers exactly this and is load-bearing (see below).

### 4. Revert-red: tests are structurally capable of catching the bug

Two independent mutations, not just a blanket revert:

**Mutation A** — drop only the `from_dict` normalization (keep `LayerResolutionError` importable so collection still works):
`13 failed, 60 passed`. Reds include `test_from_dict_normalizes_both_layer_lists`, `test_roundtrip_is_idempotent_for_name_form`, `test_soft_preference_cost_matches_the_integer_form`, and all 6 CLI preload tests.

**Mutation B** — reintroduce the `CopperLayer.from_kicad_name()` trap as a short-circuit ahead of the stack lookup:
`13 failed, 60 passed`, including every `test_b_cu_resolves_to_the_stacks_last_index` parametrization, `test_b_cu_is_not_the_copper_layer_enum_value_on_four_layers`, `test_name_absent_from_stack_raises`, and `test_auto_detected_stack_resolves_b_cu_to_three`.

The suite detects both the missing fix and the specific wrong-index-space fix. Unmutated: `73 passed`.

### 5. Scope guard: confirmed by machine

`git diff origin/main...HEAD -U0 | grep '^@@'` on `route_cmd.py` returns exactly two hunks, both post-integration line numbers inside the `--net-class-map` preload block:

```
@@ -11791,0 +11792,15 @@ def main(...)
@@ -11793   +11808,19 @@ def main(...)
```

Zero hunks in `route_cmd.py:3755-3860` (`_apply_net_class_map_sidecar` / `_apply_pairwise_clearance`), zero at `:13740-13790`, and `router/pairwise_clearance.py` is untouched — all reserved for #4588 / PR #4603, already on `main`. All new CLI code is inline in the preload block; no new module-level helper was added. Files changed: `router/rules.py`, `cli/route_cmd.py`, and two test files. Nothing else.

The stack-selection dict mirrors the `--layers` argparse choices exactly (`auto`, `2`, `4`, `4-sig`, `4-all`, `6` at `route_cmd.py:10714-10715`) — no drift.

### 6. Integration correctness on the merged tree

All evidence above was gathered against `b90d32b8`, not the old base. Semantic coexistence with the three PRs that landed during the sweep:

- **#4588 / PR #4603 (HV pairwise gate)** — the surfaces are sequential, not overlapping. Preload normalizes names to indices and stashes the map on `args._loaded_net_class_map`; `_apply_net_class_map_sidecar` later rekeys those already-normalized objects onto board net names. `router/pairwise_clearance.py` contains zero references to `avoid_layers` / `preferred_layers`, so the pairwise gate never observes a layer value in either form. The board-02 routing runs above exercised the merged path end to end with ampacity-bearing classes.
- **#4608 (manifest keys)** — `export/`, `fleet_cmd.py`; no shared surface.
- **#4609 (fresh-interpreter probe)** — `cpp_backend.py` / `build_native_cmd.py`. The one shared surface is `cpp_backend.py:1647` calling `hard_avoided_layer_indices()`; #4609 touched the extension-verification path, not that call. The native backend is what produced the positive-control census above, so the interaction is verified by execution, not just by reading.

### 7. The `Test` CI failure was a flake — evidence

The first `Test` run on this head failed. I pulled the job log:

```
FAILED tests/perf/test_pathfinder_dormant_partner_perf.py::test_dormant_partner_optimized_not_slower
  - Optimized dormant path is 1.460x slower than legacy
    (legacy_min=0.3203s, optimized_min=0.4676s). Expected ratio < 1.25
==== 1 failed, 24146 passed, 92 skipped in 1211.35s ====
```

This is a shared-runner timing gate whose own docstring records a prior false failure at 1.065x on an artifact-only PR that passed on rerun (issue #3581). It cannot be attributable to this PR, to #4608, or to #4609:

- The benchmark times `Router._is_trace_blocked` in `router/pathfinder.py` — a file this PR does not touch (changed files are `rules.py`, `route_cmd.py`, and two test files).
- The only `rules.py` symbol it uses is `DesignRules`, which occupies HEAD lines 67-714; all eleven `rules.py` hunks land outside that range.
- The fixture constructs no net class, so `hard_avoided_layer_indices()` is never reached.
- It passes locally at this exact head (`5 passed`).

**Re-ran the job at the identical head with no code change: `Test` = pass (20m52s).** All 16 checks now green. Not a defect on this PR and not a latent defect on `main`.

### Non-blocking observation (no action required for this PR)

`hard_avoided_layer_indices()` is called from `_expand_bidirectional_neighbors` (`pathfinder.py:4638`), a per-expanded-node function, and the new resolver measures ~2.5x the old `int(layer)` form (+0.32 us/call for a 2-entry avoid list). The dominant cost is the `context=f"net class {self.name!r} avoid_layers"` f-string, which is now built eagerly on every call even though it is only consumed on the error path. This only affects nets with `target_ampacity` set or `--strict-layers`, the per-node recomputation predates this PR (#4433), and Board 05 Routing Regression passes at 44m54s — so this is not a merge blocker. If it ever shows up in a profile, making the context lazy is a one-line change.

### Summary

The fix is at the right layer (normalize once at deserialization, so neither the soft cost bias nor either grid backend ever sees a name), fails loud rather than silently dropping a hard-avoid entry, correctly refuses to guess `B.Cu` without a stack, and carries tests that demonstrably catch both the original bug and the wrong-index-space trap. Approving.
